llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp

```diff
@@ -0,0 +1,2309 @@
+                                : getParentScope(SP)->getParentScope()
+                          : getParentScope(SP);
+    // The 'getParentScope' traverses the scope hierarchy and it creates
+    // the scope chain and any associated types.
+    // Check that the 'Function' is not already in the parent.
+    if (!Function->getParent())
+      Parent->addElement(Function);
+
+    getOrCreateSubprogram(Function, SP, includeMinimalInlineScopes());
+  }
+
+  return Function;
+}
+
+// DISubprogram
+LVScope *LVIRReader::getOrCreateSubprogram(LVScope *Function,
+                                           const DISubprogram *SP,
+                                           bool Minimal) {
+  assert(Function && "Invalid logical element");
+  assert(SP && "Invalid metadata node.");
+  LLVM_DEBUG({
+    dbgs() << "\n[getOrCreateSubprogram] DISubprogram\n";
+    SP->dump();
+  });
+
+  if (Function->getIsFinalized())
+    return Function;
+  Function->setIsFinalized();
+
+  // Get 'declaration' node in order to generate the DW_AT_specification.
+  if (const DISubprogram *SPDecl = SP->getDeclaration()) {
+    if (!Minimal) {
+      // Build the declaration now to ensure it precedes the definition.
+      getOrCreateSubprogram(SPDecl);
+    }
+  }
+
+  // Check for additional retained nodes.
+  for (const DINode *DN : SP->getRetainedNodes()) {
+    if (const auto *IE = dyn_cast<DIImportedEntity>(DN))
+      constructImportedEntity(Function, IE);
+    else if (const auto *TTP = dyn_cast<DITemplateTypeParameter>(DN))
+      constructTemplateTypeParameter(Function, TTP);
+    else if (const auto *TVP = dyn_cast<DITemplateValueParameter>(DN))
+      constructTemplateValueParameter(Function, TVP);
+  }
+
+  applySubprogramAttributes(Function, SP);
+
+  // Check if we are dealing with the Global Init/Cleanup Function.
+  if (SP->isArtificial() && SP->isLocalToUnit() && SP->isDefinition() &&
+      SP->getName().empty())
+    Function->setName(SP->getLinkageName());
+
+  return Function;
+}
+
+void LVIRReader::constructSubprogramArguments(LVScope *Function,
+                                              const DITypeRefArray Args) {
+  assert(Function && "Invalid logical element");
+  LLVM_DEBUG({
+    dbgs() << "\n[constructSubprogramArguments] DITypeRefArray\n";
+    for (unsigned i = 1, N = Args.size(); i < N; ++i) {
+      if (const DIType *Ty = Args[i])
+        Ty->dump();
+    }
+  });
+
+  for (unsigned I = 1, N = Args.size(); I < N; ++I) {
+    const DIType *Ty = Args[I];
+    LVElement *Parameter = nullptr;
+    if (Ty) {
+      // Create a formal parameter.
+      LVElement *ParameterType = getOrCreateType(Ty);
+      Parameter = createElement(dwarf::DW_TAG_formal_parameter);
+      if (Parameter) {
+        Parameter->setType(ParameterType);
+        if (Ty->isArtificial())
+          Parameter->setIsArtificial();
+      }
+    } else {
+      // Add an unspecified parameter.
+      Parameter = createElement(dwarf::DW_TAG_unspecified_parameters);
```

Comment:
`DW_TAG_unspecified_parameters` is used to represent variadic functions.  This looks more like 'unknown' type, but not totally sure.  Should a `DW_TAG_formal_parameter` be created instead, but without setting type information?